sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -36,75 +35,155 @@ class GpuGenerateExecSparkPlanMeta(
-      case PosExplode(Alias(Literal(_, ArrayType(_, _)), _)) => // Nothing
-      case Explode(CreateArray(_, _)) => // Nothing
-      case Explode(Literal(_, ArrayType(_, _))) => // Nothing
-      case Explode(Alias(Literal(_, ArrayType(_, _)), _)) => // Nothing
-      case _ => willNotWorkOnGpu("Only posexplode of a created array is currently supported")
-    }
-
-    if (gen.outer) {
-      willNotWorkOnGpu("outer is not currently supported")
+    if (gen.outer && !isOuterSupported(gen.generator)) {
+      willNotWorkOnGpu(s"outer is not currently supported with ${gen.generator.nodeName}")
     }
   }
 
-  /**
-   * Convert what this wraps to a GPU enabled version.
-   */
   override def convertToGpu(): GpuExec = {
     GpuGenerateExec(
-      gen.generator.isInstanceOf[PosExplode],
-      arrayExprs.map(_.convertToGpu()),
+      childExprs.head.convertToGpu().asInstanceOf[GpuGenerator],
       gen.requiredChildOutput,
+      gen.outer,
       gen.generatorOutput,
       childPlans.head.convertIfNeeded())
   }
 }
 
+object GpuGenerateExecSparkPlanMeta {
+  private val OUTER_SUPPORTED_GENERATORS: Set[String] = Set()
+
+  def isOuterSupported(gen: Generator): Boolean = {
+    OUTER_SUPPORTED_GENERATORS.contains(gen.nodeName)
+  }
+}
+
 /**
- * Takes the place of GenerateExec(PosExplode(CreateArray(_))).  It would be great to do it in a
- * more general case but because we don't support arrays/maps, we have to hard code the cases
- * where we don't actually need to put the data into an array first.
+ * GPU overrides of [[org.apache.spark.sql.catalyst.expressions.Generator]], corporate with
+ * `GpuGenerateExec`.
  */
+trait GpuGenerator extends GpuUnevaluable {
+
+  override def dataType: DataType = ArrayType(elementSchema)
+
+  override def foldable: Boolean = false
+
+  override def nullable: Boolean = false
+
+  /**
+   * The output element schema.
+   */
+  def elementSchema: StructType
+
+  /**
+   * Apply generator to produce result ColumnarBatch from input batch.
+   *
+   * This is a specialized method for GPU runtime, which is called by GpuGenerateExec who owns
+   * the generator. The reason of creating a new method rather than implementing columnarEval is
+   * that generator is an integrated Table transformer instead of column transformer in terms of
+   * cuDF.
+   *
+   * @param inputBatch projected input data, which ensures appending columns are ahead of
+   *                   generators' inputs. So, generators can distinguish them with an offset.
+   * @param generatorOffset column offset of generator's input columns in `inputBatch`
+   * @param outer when true, each input row will be output at least once, even if the output of the
+   *              given `generator` is empty.
+   *
+   * @return result ColumnarBatch
+   */
+  def generate(inputBatch: ColumnarBatch,
+    generatorOffset: Int,
+    outer: Boolean): ColumnarBatch
+}
+
+trait GpuCollectionGenerator extends GpuGenerator {
```

Comment:
I'm not sure that we need to match the Spark class hierarchy here.  Spark is doing outer/etc in a very different way and the CollectionGenerator in Spark is really there for code generation, as per the comments on that class. I think we can clean things up for now and add them back in, if we reach a situation where we need it.

Reply:
I removed this.